### PR TITLE
feat: celo api

### PR DESCRIPTION
**Pull request by dextracker**
# Description

# Testing Instructions

# Checklist

-   [ ] Update [documentation](https://github.com/0xProject/website/blob/development/mdx/api/index.mdx) as needed. **Website Documentation PR:**
-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Test changes on the staging environment with [Matcha API Staging](https://api-staging.matcha.xyz).

    -   [ ] SRA/Limit orders
    -   [ ] Swap endpoints
    -   [ ] Meta transaction endpoints
    -   [ ] Depth charts

    For more information see `0x API Matcha smoke test runbook` in Quip.

### CODE REVIEW

File: .env_example [hunk 1]
```diff
@@ -1,44 +0,0 @@
-# Required
```
Comment (dekz) [inline .env_example:1]: We keep this for documenting the ENV vars for others

File: hardhat.config.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
+require("@nomiclabs/hardhat-waffle");
```
Comment (dekz) [inline hardhat.config.js:1]: Not needed to be checked in, perhaps look at a `.gitignore` for this file if you want to keep it around locally

File: package.json [hunk 3]
```diff
@@ -140,20 +140,22 @@
...
         "@0x/quote-server": "^7.0.2",
         "@0x/subproviders": "~6.5.4",
-        "@0x/token-metadata": "^0.1.6",
+        "@0x/token-metadata": "^0.1.8",
         "@0x/types": "^3.3.4",
         "@0x/utils": "^6.4.4",
         "@0x/web3-wrapper": "^7.6.1",
         "@ethersproject/hdnode": "^5.2.0",
+        "@nomiclabs/hardhat-ethers": "^2.0.0",
+        "@nomiclabs/hardhat-waffle": "^2.0.1",
```
Comment (dekz) [inline package.json:158]:
```suggestion

```

File: package.json [hunk 4]
```diff
@@ -162,6 +164,7 @@
         "dotenv": "^8.1.0",
         "elastic-apm-node": "^3.7.0",
         "ethereum-types": "^3.5.0",
+        "ethereum-waffle": "^3.2.0",
```
Comment (dekz) [inline package.json:167]:
```suggestion
```

File: src/services/swap_service.ts [hunk 5]
```diff
@@ -217,6 +217,8 @@ export class SwapService {
 
         // Check if integrator ID specifically whitelists a set of maker URIs. If whitelist is "undefined" then it
         // means all integrators will be enabled.
+
+        const CELO_CHAIN_ID = 42220;
```
Comment (dekz) [inline src/services/swap_service.ts:221]:
```suggestion

```

Should be able to use `ChainId.Celo` instead

File: src/services/swap_service.ts [hunk 6]
```diff
@@ -365,35 +366,66 @@ export class SwapService {
...
-            decodedUniqueId,
-            sellTokenToEthRate,
-            buyTokenToEthRate,
-            quoteReport,
-            extendedQuoteReportSources,
-            priceComparisonsReport,
-        };
+        let apiSwapQuote: GetSwapQuoteResponse;
+        switch (CHAIN_ID) {
+            case CELO_CHAIN_ID:
```
Comment (dekz) [inline src/services/swap_service.ts:371]:
```suggestion
            case ChainId.Celo:
```

File: src/services/swap_service.ts [hunk 7]
```diff
@@ -365,35 +366,66 @@ export class SwapService {
...
+                    data,
+                    value: adjustedValue,
+                    gas: worstCaseGasEstimate,
+                    estimatedGas: conservativeBestCaseGasEstimate,
+                    from: takerAddress,
+                    gasPrice,
+                    protocolFee,
+                    minimumProtocolFee: BigNumber.min(protocolFee, bestCaseProtocolFee),
+                    // NOTE: Internally all ETH trades are for WETH, we just wrap/unwrap automatically
+                    buyTokenAddress: isETHBuy ? '0x471ece3750da237f93b8e339c536989b8978a438' : buyToken,
```
Comment (dekz) [inline src/services/swap_service.ts:386]: For a fresh reader they'd have no idea why this is duplicated with only some (2?) lines of difference. Leave a comment for future self why this is needed.

It's probably cleaner to have a helper function which takes the chain and returns the Native token address

Something like


```typescript
function nativeAssetForChain(chainId: ChainId): string => {
    switch (chainId) {
        case ChainId.Celo:
            return '0x471ece3750da237f93b8e339c536989b8978a438';
        default:
            return ETH_TOKEN_ADDRESS;
    }
}
```

**Review by dekz — CHANGES_REQUESTED** [review]

File: src/services/swap_service.ts [hunk 8]
```diff
@@ -282,8 +284,7 @@ export class SwapService {
             protocolFeeInWeiAmount: bestCaseProtocolFee,
         } = swapQuote.bestCaseQuoteInfo;
         const { protocolFeeInWeiAmount: protocolFee, gas: worstCaseGas } = swapQuote.worstCaseQuoteInfo;
-        const { gasPrice, sourceBreakdown, quoteReport, extendedQuoteReportSources, priceComparisonsReport } =
```
Comment (kimpers) [inline src/services/swap_service.ts:285]: Is it intentional to remove `extendedQuoteReportSources`?

File: src/services/swap_service.ts [hunk 9]
```diff
@@ -381,7 +384,7 @@ export class SwapService {
             minimumProtocolFee: BigNumber.min(protocolFee, bestCaseProtocolFee),
             // NOTE: Internally all ETH trades are for WETH, we just wrap/unwrap automatically
             buyTokenAddress: isETHBuy ? ETH_TOKEN_ADDRESS : buyToken,
-            sellTokenAddress: isETHSell ? ETH_TOKEN_ADDRESS : sellToken,
```
Comment (dorothy-zbornak) [inline src/services/swap_service.ts:384]: I think we want the old behavior?

**Review by dorothy-zbornak — APPROVED** [review]

**Pull request merged**